src/storage.rs

```diff
@@ -0,0 +1,43 @@
+use std::sync::Arc;
+
+const MAX_INLINE: usize = 39;
+
+#[derive(Clone)]
+pub enum Storage {
+    /// hash is stored inline. if it is smaller than 39 bytes it should be padded with 0u8
+    Inline([u8; MAX_INLINE]),
+    /// hash is stored on the heap. this must be only used if the hash is actually larger than
+    /// 39 bytes to ensure an unique representation.
+    Heap(Arc<[u8]>),
+}
+
+impl Storage {
+    /// The raw bytes. Note that this can be longer than the data this storage has been created from.
+    pub fn bytes(&self) -> &[u8] {
+        match self {
+            Storage::Inline(bytes) => bytes,
+            Storage::Heap(data) => &data,
+        }
+    }
+
+    /// creates storage from a vec. Note that this will not preserve the size.
+    pub fn copy_from_slice(slice: &[u8]) -> Self {
```

Comment:
I think the `copy_from_slice` overwrites existing slices, so you want `from_slice` here